hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -30,125 +30,189 @@
-    def failures(self):
-        return self.runs - self.successes
-
-    @property
-    def name(self):
-        return self.pass_function.__name__
-
-    def __eq__(self, other):
-        return self.index == other.index
-
-    def __hash__(self):
-        return hash(self.index)
-
-    def __lt__(self, other):
-        return self.key() < other.key()
-
-    def key(self):
-        # Smaller is better.
-        return (self.runs, self.failures, self.calls, self.index)
+    return (len(buffer), buffer)
 
 
 class Shrinker(object):
     """A shrinker is a child object of a ConjectureRunner which is designed to
-    manage the associated state of a particular shrink problem.
+    manage the associated state of a particular shrink problem. That is, we
+    have some initial ConjectureData object and some property of interest
+    that it satisfies, and we want to find a ConjectureData object with a
+    shortlex (see sort_key above) smaller buffer that exhibits the same
+    property.
+
+    Currently the only property of interest we use is that the status is
+    INTERESTING and the interesting_origin takes on some fixed value, but we
+    may potentially be interested in other use cases later.
+    However we assume that data with a status < VALID never satisfies the predicate.
+
+    The shrinker keeps track of a value shrink_target which represents the
+    current best known ConjectureData object satisfying the predicate.
+    It refines this value by repeatedly running *shrink passes*, which are
+    methods that perform a series of transformations to the current shrink_target
+    and evaluate the underlying test function to find new ConjectureData
+    objects. If any of these satisfy the predicate, the shrink_target
+    is updated automatically. Shrinking runs until no shrink pass can
+    improve the shrink_target, at which point it stops. It may also be
+    terminated if the underlying engine throws RunIsComplete, but that
+    is handled by the calling code rather than the Shrinker.
+
+    =======================
+    Designing Shrink Passes
+    =======================
+
+    Generally a shrink pass is just any function that calls
+    cached_test_function and/or incorporate_new_buffer a number of times,
+    but there are a couple of useful things to bear in mind.
+
+    The main invariant that shrink passes should satisfy is that if
+    running a shrink pass failed to make progress, running it again
+    should still fail to make progress. This means that it is
+    fine to include non-determinism in the design of a shrink pass
+    (we do in several), but until the shrink pass has successfully
+    made progress that non-determinism should only affect the order
+    in which things are tried. e.g. it's fine to try each of N deletions
+    or replacements in a random order, but it's not OK to try N random
+    deletions (unless you have already shrunk at least once, though we
+    don't currently take advantage of this loophole).
```

Comment:
Might be worth noting the idea behind this restriction: The shrinker wants to be able to assume that a failed pass is at fixed-point, and avoid running it again until some other pass has made progress.